File: 3sem/4lab/Banks/BankAccounts/DebitBankAccount.cs
using Banks.Clocks;
using Banks.Entities;
using Banks.Exceptions;
using Banks.Models.BankConfigurations;

namespace Banks.BankAccounts;

public class DebitBankAccount : IBankAccount
{
    private decimal _dailyPercents;

    public DebitBankAccount(Bank bank, Client client, IClock clock)
    {
        ArgumentNullException.ThrowIfNull(bank);
        ArgumentNullException.ThrowIfNull(client);
        ArgumentNullException.ThrowIfNull(clock);

        Balance = 0;
        _dailyPercents = 0;

        Client = client;
        Configuration = bank.Configuration.DebitConfiguration;

        Id = Guid.NewGuid();

        clock.DayChanged += OnDayChanged;
        clock.MonthChanged += OnMonthChanged;
    }

    public Guid Id { get; }
    public Client Client { get; }
    public DebitBankAccountConfiguration Configuration { get; }
    public decimal Balance { get; private set; }

    public bool CanDeposit(decimal moneyAmount)
    {
        return Client.IsVerified() || moneyAmount <= Configuration.UnverifiedLimit;
    }

    public bool CanWithdraw(decimal moneyAmount)
    {
        if (!Client.IsVerified() && moneyAmount > Configuration.UnverifiedLimit)
            return false;

        return moneyAmount <= Balance;
    }

    public void Deposit(decimal moneyAmount)
    {
        if (moneyAmount < 0)
            throw InvalidBankAccountOperationException.OnInvalidMoneyAmount();

        Balance += moneyAmount;
    }

    public void Withdraw(decimal moneyAmount)
    {
        if (moneyAmount < 0)
            throw InvalidBankAccountOperationException.OnInvalidMoneyAmount();

        if (!CanWithdraw(moneyAmount))
            throw InvalidBankAccountOperationException.OnShortageOfMoney(moneyAmount, Balance);

        Balance -= moneyAmount;
    }

    private void OnDayChanged()
    {
        _dailyPercents += Balance * Configuration.InterestRate / 365;
    }

    private void OnMonthChanged()
    {
        Balance += _dailyPercents;
        _dailyPercents = 0;
    }
}

File: 4sem/1lab/src/main/java/BanksRemake/Domain/BankAccounts/DebitBankAccount.java
package BanksRemake.Domain.BankAccounts;

import BanksRemake.Domain.Clocks.Clock;
import BanksRemake.Domain.Clocks.ClockSubscriber;
import BanksRemake.Domain.Entities.Bank;
import BanksRemake.Domain.Entities.Client;
import BanksRemake.Domain.Exceptions.InvalidBankAccountOperationException;
import BanksRemake.Domain.Models.BankConfigurations.DebitBankAccountConfiguration;

import java.util.UUID;

/**
 * Bank account that earns interest every month.
 */
public class DebitBankAccount implements BankAccount, ClockSubscriber {
    private final UUID id;
    private final Client client;
    private final DebitBankAccountConfiguration configuration;
    private double balance;
    private double dailyPercents;

    public DebitBankAccount(Bank bank, Client client, Clock clock)  {
        this.balance = 0;
        this.dailyPercents = 0;

        this.client = client;
        this.configuration = bank.getConfiguration().debitBankAccountConfiguration();

        id = UUID.randomUUID();

        clock.addSubscriber(this);
    }

    @Override
    public UUID getId() {
        return id;
    }

    @Override
    public double getBalance() {
        return balance;
    }

    @Override
    public Client getClient() {
        return client;
    }

    @Override
    public boolean canDeposit(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        return client.isVerified() || moneyAmount <= configuration.unverifiedLimit();
    }

    @Override
    public boolean canWithdraw(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        if (!client.isVerified() && moneyAmount > configuration.unverifiedLimit()) {
            return false;
        }

        return moneyAmount <= balance;
    }

    @Override
    public void deposit(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        balance += moneyAmount;
    }

    @Override
    public void withdraw(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        if (!canWithdraw(moneyAmount)) {
            throw InvalidBankAccountOperationException.onShortageOfMoney(moneyAmount, balance);
        }

        balance -= moneyAmount;
    }

    @Override
    public void reactOnDayChanged() {
        dailyPercents += balance * configuration.interestRate() / 365;
    }

    @Override
    public void reactOnMonthChanged() {
        balance += dailyPercents;
        dailyPercents = 0;
    }
}
